### ISSUE TITLE
Repository

### ISSUE THREAD

**Issue by glatard** [boutiques/boutiques#34]
Design a repository format to store Boutiques descriptors. Repository is based on git so that version/author tracking and access permissions are supported by default. Repository could include metadata that contain:
- Description of the repository. 
- Location of files in repository
- Tool groups (e.g. "FSL tools")
- Dependencies between tools (e.g. "Niakfmripreprocess" requires "psomworker")
- Mirrors / other known repositories
- Signatures / checksums (maybe it's not required in a git repository?)

**Comment by glatard** [comment]
Could also add DOIs to Boutiques tools

**Comment by glatard** [comment]
Check that tool names are unique in the repo

**Comment by glatard** [comment]
CBRAIN could create plugins from descriptors in the repo.

**Comment by glatard** [comment]
Or maybe the boutiques repo should be a federation of git repos containing boutiques descriptors.

**Comment by glatard** [comment]
Perhaps make it just an index (in Neurolinks?) augmented with functionalities like validation, etc

**Comment by glatard** [comment]
`bosh-publish` can now publish descriptors to (Neurolinks)[https://brainhack101.github.io/neurolinks]. Not sure if we still want to do this.

**Comment by glatard** [comment]
brep is a prootypte of a git repo containing all the boutiques descriptors as sub-modules : https://github.com/glatard/brep/tree.

Suggesetion from the meeting: explore git-annex to avoid having to downalod a complete sub-moudle just to get the JSON boutiques descriptor from it.

**Comment by glatard** [comment]
data lad may help here.

**Comment by glatard** [comment]
Suggestion from GigaScience reviewer:
```
Furthermore, the feature of `bosh` to upload descriptors to a common repository (i.e., NeuroLinks) is an important one. While links to descriptors are certainly useful, I would like to suggest a common repository of the descriptor files themselves. Such a repository would reduce redundancy in coding effort because contributors could see which programs already have descriptors. Upstream software maintainers could maintain their Boutiques descriptors in such a repository. The burden on new users might also be reduced because they would have access to all Boutiques descriptors in a single location. If a common repository like this were in place, `bosh` might be extended to accept descriptors in a more abstract form. For example, instead of specifying a path to an existing descriptor file, a user could specify "fsl.bet.v1", and `bosh` could pull the relevant descriptor from the common repository. A similar effect might be achieved using the existing JSON file that lists the
tools on the NeuroLinks website.
```

**Comment by glatard** [comment]
We need to move on with this. Here is a list of desired features for the repo:

1. Accessible offline, to facilitate tool (pipeline) development and testing. 
2. Easy to install.
3. Searchable through `bosh search`. 

Current plan:
* create a `brep` repo under `boutiques`, where descriptors would be committed.
* add commands in bosh to clone the repo locally and search in it. Update `bosh publish` to publish in it and Neurolinks to refer to it. `bosh search` and `bosh publish` could clone the repo locally if not available already. 
* `bosh exec` could also work on a tool name from the repo instead of a descriptor path, as suggested by the reviewer.

**Comment by glatard** [comment]
From discussion with @gkiar today:
* `brep` sounds like an overkill for the following reasons:
   * versioning and releases are problematic because users may want to use different versions of different tools, not to mention packages.
   * it keeps descriptors far away from the code (see previous discussions)

Our current idea is to leverage Neurolinks as follows:
   1. make Boutiques entries in neurolinks immutable: (a) modify bosh publish to commit a URL of a particular commit of the descriptor instead of the latest version (b) add the version number of the identifier.
   2. implement:
      -  `bosh search` that greps in https://raw.githubusercontent.com/brainhack101/neurolinks/master/jsons/tool.json and returns a table containing the identifiers and (truncated) descriptions of the tools 
     - `bosh get` that downloads a descriptor based on its neurolinks identifier.

**Comment by shots47s** [comment]
Well, I think what you are talking about should have some form of ability
to tag by user.  The reason why I say that is that nice in docker hub and
other such tools to have "my" set.  Also might increase adoption as some
form of attribution.

But saying that, i am cool with what you have outlined.
-- 
Shawn T. Brown, PhD.
Chief Software Architect / CBRAIN
McGill Centre for Integrative Neuroscience (MCIN)
McGill University
Montreal, QC, Canada
Email: stbrown@mcin.ca
Phone: 514-237-2960 <(514)%20237-2960>

On Jan 26, 2018, at 5:57 PM, Tristan Glatard <notifications@github.com>
wrote:

From discussion with @gkiar <https://github.com/gkiar> today:

   - brep sounds like an overkill for the following reasons:
      - versioning and releases are problematic because users may want to
      use different versions of different tools, not to mention packages.
      - it keeps descriptors far away from the code (see previous
      discussions)

Our current idea is to leverage Neurolinks as follows:

   1. make Boutiques entries in Neurolinks immutable:
   (a) modify bosh publish to commit a URL of a particular commit of the
   descriptor instead of the latest version.
   (b) add the version number of the identifier.
   (c) write guidelines to accept PRs about Boutiques descriptors in
   Neurolinks.
   2. implement:
   (a) bosh search that greps in https://raw.githubusercontent.
   com/brainhack101/neurolinks/master/jsons/tool.json
   <https://raw.githubusercontent.com/brainhack101/neurolinks/master/jsons/tool.json>
   and returns a table containing the identifiers and (truncated) descriptions
   of the Boutiques tools.
   (b) bosh get that downloads a descriptor based on its neurolinks
   identifier.
   (c) bosh launch <simulate|exec> <neurolinks_id> that calls bosh get to
   get the descriptor and launch the tool.

@camarasu <https://github.com/camarasu> @shots47s
<https://github.com/shots47s> what do you think?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/boutiques/boutiques/issues/34#issuecomment-360922321>,
or mute the thread

**Comment by gkiar** [comment]
@shots47s in the `tool.json` being grepped, there is a field called "author," so people could search by the contributor such as Dockerhub. Is this what you mean?

**Comment by glatard** [comment]
the author field could also be shown in the table printed by `bosh search`.

**Comment by shots47s** [comment]
Yes on both.

**Comment by camarasu** [comment]
As discussed with Tristan, I find it a good idea to leverage Neurolinks as you suggest. 
However, I think we should also consider 'non-neuro' applications. One solution would be to (i) re-use the Neurolinks model for other application domains and (ii) have a top-level Boutiques catalog listing all  existing repositories. 
This can be done when really needed, doesn't have to be in place immediately for the paper resubmission.

**Comment by gkiar** [comment]
I agree! We should also put testing data next to the descriptor when `bosh publish` is run.

**Comment by glatard** [comment]
For now test data will have to be in the tool container so that's not an issue.

**Comment by shots47s** [comment]
Potential use case for DataLad?

-- 
Shawn T. Brown, PhD.
Chief Software Architect / CBRAIN
McGill Centre for Integrative Neuroscience (MCIN)
McGill University
Montreal, QC, Canada
Email: stbrown@mcin.ca
Phone: 514-237-2960

**Comment by glatard** [comment]
Closing this issue as it is implemented in #240. Will open another one for search functionallity.
